metric/id/m3/id.go

```diff
@@ -0,0 +1,197 @@
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+// Package m3 describes m3 metric id information.
+//
+// Each M3 metric id contains a metric name and a set of tag pairs. In particular,
+// it conforms to the following format:
+//
+// m3+<metric_name>+tagName1=tagValue1,tagName2=tagValue2,...tagNameN=tagValueN,
+//
+// Where the tag names are sorted alphabetically in ascending order.
+//
+// An example m3 metrid id is as follows:
+// m3+response_code+env=bar,service=foo,status=404,type=counters
+package m3
+
+import (
+	"bytes"
+	"fmt"
+	"sort"
+
+	"github.com/m3db/m3metrics/metric/id"
+)
+
+const (
+	componentSplitter = '+'
+	tagNameSplitter   = '='
+	tagPairSplitter   = ','
+)
+
+var (
+	errInvalidM3Metric = fmt.Errorf("invalid m3 metric")
+	m3Prefix           = []byte("m3+")
+	nameTagName        = []byte("name")
+	rollupTagPair      = id.TagPair{
+		Name:  []byte("m3_rollup"),
+		Value: []byte("true"),
+	}
+)
+
+// NewRollupID generates a new rollup id given the new metric name
+// and a list of tag pairs. Note that tagPairs are mutated in place.
+func NewRollupID(name []byte, tagPairs []id.TagPair) []byte {
+	var buf bytes.Buffer
+
+	// Adding rollup tag pair to the list of tag pairs.
+	tagPairs = append(tagPairs, rollupTagPair)
+	sort.Sort(id.TagPairsByNameAsc(tagPairs))
+
+	buf.Write(m3Prefix)
+	buf.Write(name)
+	buf.WriteByte(componentSplitter)
+	for i, p := range tagPairs {
+		buf.Write(p.Name)
+		buf.WriteByte(tagNameSplitter)
+		buf.Write(p.Value)
+		if i < len(tagPairs)-1 {
+			buf.WriteByte(tagPairSplitter)
+		}
+	}
+
+	return buf.Bytes()
+}
+
+// TODO(xichen): pool the mids.
+type mid struct {
+	id     []byte
+	iterFn id.SortedTagIteratorFn
+}
+
+// NewID creates a new m3 metric id.
+func NewID(id []byte, iterFn id.SortedTagIteratorFn) id.ID {
+	return mid{id: id, iterFn: iterFn}
+}
+
+func (id mid) Bytes() []byte { return id.id }
+
+func (id mid) TagValue(tagName []byte) ([]byte, bool) {
+	_, tagPairs, err := NameAndTags(id.Bytes())
```

Comment:
Do we want to call `NameAndTags` here, seems like it might add a bit of overhead. For example, if we're interested in the value for one of the first tags, we will have to iterate over the whole ID in `NameAndTags` even though the we are interested in is in the beginning of the id.